### ISSUE TITLE
Group-by and aggregation operation on multiple columns can produce incorrect output

### ISSUE THREAD

**Issue by wicknicks** [confluentinc/ksql#6101]
**Describe the bug**
Given the following group-by query: 

```
CREATE TABLE gb_out WITH (kafka_topic='gb_out', value_format='json') 
AS select a, b, sum(c) 
FROM test GROUP BY a, b EMIT CHANGES;
```

where `test` is declared as: 

```
CREATE STREAM test (a varchar, b varchar, c double)
WITH (kafka_topic='test', value_format='json', partitions=1);
```

KSQL will create keys by concatenating columns `a` and `b` using `|+|` as a delimiter. If this delimiter appears in the string, then the final key might appear identical to very different values in the original columns, creating incorrect aggregations. 

**To Reproduce**
Start CLI, run ksql-server v0.11, create the above queries, and produce the following records. The examples here use `kafkacat`, but any utility can be used:

```
echo '{"a": "goog|+|usd", "b": "usd", "c": 120.50}' | kafkacat -P -b localhost:29092 -t test
echo '{"a": "goog", "b": "usd|+|usd", "c": 100.00}' | kafkacat -P -b localhost:29092 -t test
```

Now, check the results: 

```
ksql> select * from gb_out emit changes;
+---------------------------------+---------------------+
|KSQL_COL_1                       |KSQL_COL_0           |
+---------------------------------+---------------------+
|goog|+|usd|+|usd                 |120.5                |
|goog|+|usd|+|usd                 |220.5                |
+---------------------------------+---------------------+
```

**Expected behavior**
The second row in CLI output above is incorrect. Instead of 220.5, we should just have seen a second row with the value 100.0.

**Actual behaviour**
Two columns are incorrectly combined to form an identical group-by key accumulating more values than what they should have.

A solution could be to prefix the hash of the composite-key map to the group-by-key (in `KSQL_COL_1` above) to distinguish between records to get around such cases. In our example above, `hash({"a": "goog|+|usd", "b": "usd"})` will be different from `hash({"a": "goog", "b": "usd|+|usd"})`.

**Additional context**
KSQL Server version: 0.11

**Comment by agavra** [comment]
Thanks for reporting this @wicknicks! 

I think the structural plan to solve this issue is to use structured keys (see #5603 for example, I don't know if we have a more specific ticket for structured keys than that). Your hash suggestion would work, alternatively (perhaps a more user-friendly approach) would be to just allow users to specify the delimiter manually in case they suspect that their content will have `|+|` in it. Backwards compatibility, obviously, would need to be kept in mind in both cases.

I'm going to triage this as lower priority (P1) since I suspect this bug is rather rare as we haven't run into the complaint yet and we have a structural way of solving this.

**Comment by wicknicks** [comment]
In the next release, we should at least disallow such queries. if multi-key group by always need delimiters to be specified, then I think KSQL should not allow creating queries that can give incorrect answers. in a long running query, incorrect answers might creep in and users might not notice it, until its too late.

**Comment by agavra** [comment]
Closing this since any new queries that have non-complex keys will require changing the format to a multi-key format. The bug will no longer be applicable to any newly issued queries (though for compatibility, any query issued before the multi-column keys were introduced will remain with this possibility)
